checklist/http.go

```diff
@@ -0,0 +1,201 @@
+package checklist
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/matryer/way"
+	"github.com/pkg/errors"
+	"go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp"
+
+	"github.com/jarri-abidi/todo"
+)
+
+func NewServer(service Service, logger log.Logger) http.Handler {
+	s := server{service: service}
+
+	var handleSaveTask http.Handler
+	handleSaveTask = s.handleSaveTask()
+	handleSaveTask = httpLoggingMiddleware(logger, "handleSaveTask")(handleSaveTask)
+	handleSaveTask = otelhttp.NewHandler(handleSaveTask, "handleSaveTask")
+	// wire other middlewares here
+
+	var handleListTasks http.Handler
+	handleListTasks = s.handleListTasks()
+	handleListTasks = httpLoggingMiddleware(logger, "handleListTasks")(handleListTasks)
+	handleListTasks = otelhttp.NewHandler(handleListTasks, "handleListTasks")
+
+	var handleRemoveTask http.Handler
+	handleRemoveTask = s.handleRemoveTask()
+	handleRemoveTask = httpLoggingMiddleware(logger, "handleRemoveTask")(handleRemoveTask)
+	handleRemoveTask = otelhttp.NewHandler(handleRemoveTask, "handleRemoveTask")
+
+	var handleToggleTask http.Handler
+	handleToggleTask = s.handleToggleTask()
+	handleToggleTask = httpLoggingMiddleware(logger, "handleToggleTask")(handleToggleTask)
+	handleToggleTask = otelhttp.NewHandler(handleToggleTask, "handleToggleTask")
+
+	router := way.NewRouter()
+
+	router.Handle("POST", "/checklist/v1/tasks", handleSaveTask)
+	router.Handle("GET", "/checklist/v1/tasks", handleListTasks)
+	router.Handle("DELETE", "/checklist/v1/task/:id", handleRemoveTask)
+	router.Handle("PATCH", "/checklist/v1/task/:id", handleToggleTask)
+
+	router.NotFound = http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) { writeError(w, ErrResourceNotFound) })
+
+	return router
+}
+
+const (
+	contentTypeKey   = "Content-Type"
+	contentTypeValue = "application/json; charset=utf-8"
+)
+
+var (
+	ErrNonNumericTaskID = errors.New("task id in path must be numeric")
+	ErrResourceNotFound = errors.New("resource not found")
+	ErrMethodNotAllowed = errors.New("method not allowed")
+)
+
+type ErrInvalidRequestBody struct{ err error }
+
+func (e ErrInvalidRequestBody) Error() string { return fmt.Sprintf("invalid request body: %v", e.err) }
+
+type server struct {
+	service Service
+}
+
+func (s *server) handleSaveTask() http.HandlerFunc {
+	type request struct {
+		Name string `json:"name"`
+	}
+	type response struct {
```

Comment:
don't you think this struct is getting duplicated on `line 100` too?

Reply:
its a good thing that every handler has its own request and response structs, but it is also leading to duplication of code.